pkg/auth/remote_authenticator/authenticator.go

```diff
@@ -0,0 +1,164 @@
+	"github.com/treeverse/lakefs/pkg/logging"
+)
+
+const RemoteAuthSource = "remote_authenticator"
+
+// Request object that will be sent to the remote authenticator service as JSON payload in a POST request
+type AuthenticationRequest struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+}
+
+// Expected response from the remote authenticator service
+type AuthenticationResponse struct {
+	// optional, if returned then the user will be used as the official username in lakeFS
+	ExternalUserIdentifier string `json:"external_user_identifier"`
+}
+
+// RemoteAuthenticator client
+type RemoteAuthenticator struct {
+	AuthService auth.Service
+	Logger      logging.Logger
+	Config      *config.RemoteAuthenticator
+	authURL     string
+	c           *http.Client
+}
+
+func NewRemoteAuthenticator(conf *config.RemoteAuthenticator, authService auth.Service, logger logging.Logger) (auth.Authenticator, error) {
+	authUrl, err := url.JoinPath(conf.BaseURL, conf.AuthEndpoint)
+	if err != nil {
+		return nil, err
+	}
+	logger.Debugf("initializing remote authenticator with auth url %s", authUrl)
+	return &RemoteAuthenticator{
+		Logger:      logger,
+		Config:      conf,
+		authURL:     authUrl,
+		AuthService: authService,
+	}, nil
+}
+
+func (ra *RemoteAuthenticator) client() *http.Client {
+	if ra.c == nil {
+		ra.c = http.DefaultClient
+	}
+	return ra.c
+}
+
+func (ra *RemoteAuthenticator) newRequest(ctx context.Context, username, password string) (*http.Request, error) {
+	payload, err := json.Marshal(&AuthenticationRequest{Username: username, Password: password})
+
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequestWithContext(ctx, "POST", ra.authURL, bytes.NewBuffer(payload))
+	return req, err
+}
+
+func (ra *RemoteAuthenticator) doRequest(logger logging.Logger, req *http.Request) ([]byte, error) {
+	client := ra.client()
+	log := logger.WithField("url", req.URL.String())
+	log.Debug("starting http request to remote authenticator")
+
+	req.Header.Set("Content-Type", "application/json")
+	resp, err := client.Do(req)
+
+	if err != nil {
+		log.WithError(err).Debug("failed sending request to remote authenticator")
+		return nil, err
+	}
+	log = log.WithField("status_code", resp.StatusCode)
+	log.Debug("got response from remote authenticator")
+
+	defer resp.Body.Close()
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		log.WithError(err).Debug("failed reading response body")
+		return nil, err
+	}
+	return body, nil
+}
+
+func (ra *RemoteAuthenticator) AuthenticateUser(ctx context.Context, username, password string) (string, error) {
+	log := ra.Logger.WithField("input_username", username)
+
+	req, err := ra.newRequest(ctx, username, password)
+
+	if err != nil {
+		log.WithError(err).Error("failed creating request")
+		return "", auth.ErrInvalidRequest
+	}
+
+	data, err := ra.doRequest(log, req)
+
+	if err != nil {
+		return "", fmt.Errorf("doing http request %s: %w", username, err)
```

Comment:
```suggestion
		return "", fmt.Errorf("remote authenticate %s: %w", username, err)
```

Reply:
I think it's less indicative since it's less explicit on which step failed. there are many authenticators running (chain) with errors chaining when they occur. 
I prefer to keep it (specifically the part about http)